src/Subscriber/Response/PaymentMethodRouteResponseSubscriber.php

```diff
@@ -0,0 +1,182 @@
+    private $paymentMethodsService;
+
+    /**
+     * @var PaymentMethodsFilterService
+     */
+    private $paymentMethodsFilterService;
+
+    /**
+     * @var array
+     */
+    private $paymentMethodsResponse;
+
+    public function __construct(
+        AdyenPluginProvider $adyenPluginProvider,
+        PaymentMethodsService $paymentMethodsService,
+        PaymentMethodsFilterService $paymentMethodsFilterService
+    ) {
+        $this->adyenPluginProvider = $adyenPluginProvider;
+        $this->paymentMethodsService = $paymentMethodsService;
+        $this->paymentMethodsFilterService = $paymentMethodsFilterService;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            // Ensure event is executed before "Shopware\Core\System\SalesChannel\Api\StoreApiResponseListener".
+            KernelEvents::RESPONSE => ['adjustPaymentMethodsResponse', 11000],
+        ];
+    }
+
+    public function adjustPaymentMethodsResponse(ResponseEvent $event): void
+    {
+        $response = $event->getResponse();
+        if (!$response instanceof PaymentMethodRouteResponse) {
+            return;
+        }
+
+        $context = $this->getSalesChannelContext($event->getRequest());
+        if (null === $context) {
+            return;
+        }
+
+        $this->filterPaymentMethods($context, $event);
+        $this->extendPaymentMethodsData($context, $response);
+    }
+
+    public function reset()
+    {
+        $this->paymentMethodsResponse = null;
+    }
+
+    private function filterPaymentMethods(SalesChannelContext $context, ResponseEvent $event): void
+    {
+        if (true !== $event->getRequest()->query->getBoolean('onlyAvailable', false)) {
+            return;
+        }
+
+        $methods = $this->paymentMethodsFilterService->filterShopwarePaymentMethods(
+            $event->getResponse()->getPaymentMethods(),
+            $context,
+            $this->adyenPluginProvider->getAdyenPluginId()
+        );
+
+        $response = new PaymentMethodRouteResponse($methods);
+        $event->setResponse($response);
+    }
+
+    private function extendPaymentMethodsData(SalesChannelContext $context, PaymentMethodRouteResponse $response): void
+    {
+        $methods = $response->getPaymentMethods();
+        foreach ($methods as $method) {
+            if ($method->getPluginId() !== $this->adyenPluginProvider->getAdyenPluginId()) {
+                continue;
+            }
+
+            $type = $this->getPaymentMethodType($method);
+
+            $extension = new AdyenPaymentMethodDataStruct();
+            
+            $extension->setType($type);
+
+            if (!empty($type)) {
+                $extension->setPaymentMethodsResponse($this->getPaymentMethodsResponseForType($context, $type));
+            }
+
+            $method->addExtension('adyenData', $extension);
+        }
+    }
+
+    private function getPaymentMethodsResponseForType(SalesChannelContext $context, string $type)
+    {
```

Comment:
@AndreasA I want to propose a revision for this method. Instead of returning the entire $paymentMethodsResponse, we could just return the data for a single payment method (ie `$paymentMethodsResponse['paymentMethods'][0]`) and return null if the payment method data is not found.
The current implementation returns the following:
```
...
"extensions": {
    ...
    "adyenData": {
        "type": "paypal",
        "paymentMethodsResponse": {
            "paymentMethods": {
                "1": {
                    "configuration": {
                        "merchantId": "MERCHANT_ACCOUNT",
                        "intent": "capture"
                    },
                    "name": "PayPal",
                    "type": "paypal"
                }
            }
        },
        "apiAlias": "adyen_shopware_struct_adyen_payment_method_data_struct"
    }
},
```
and I think we can instead change it to the following:
```
...
"extensions": {
    ...
    "adyenData": {
        "type": "paypal",
        "paymentMethodsResponse":  {
            "configuration": {
                "merchantId": "MERCHANT_ACCOUNT",
                "intent": "capture"
            },
            "name": "PayPal",
            "type": "paypal"
        },
        "apiAlias": "adyen_shopware_struct_adyen_payment_method_data_struct"
    }
},
```
What do you think? Will this have a big impact on your existing implementation?

Reply:
@peterojo Don't we already limit the payment methods response to valid payment methods for that payment method?

And couldn't this still be an array as e.g. credit cards would have visa/mastercard etc.? or is this still just one?

Reply:
@peterojo 
```
        $paymentMethodsResponse['paymentMethods'] = array_filter(
            $paymentMethodsResponse['paymentMethods'],
            function ($value) use ($type) {
                return ($value['type'] ?? null) == $type;
            }
        );
```

However, in case you are sure that there is always only ONE payment method valid per payment method then it could be refactored. 

However, doesn't adyen require the full payment methods response when creating the component? which is why with the current implementation one can just use
```
 paymentMethodsResponse: paymentMethod.extensions.adyenData.paymentMethodsResponse, 
```

However, in theory we can change this but not sure, if we then have all the necessary data for the web component call ?

Reply:
@peterojo 

1.) Don't we already limit the payment methods response to valid payment methods for that payment method?

And couldn't this still be an array as e.g. credit cards would have visa/mastercard etc.? or is this still just one?

It should be done here:
```
        $paymentMethodsResponse['paymentMethods'] = array_filter(
            $paymentMethodsResponse['paymentMethods'],
            function ($value) use ($type) {
                return ($value['type'] ?? null) == $type;
            }
        );
```


2.) In case you are sure that there is always only ONE payment method valid per payment method then it could be refactored. 

3.) However, doesn't adyen require the full payment methods response when creating the component? which is why with the current implementation one can just use
```
 paymentMethodsResponse: paymentMethod.extensions.adyenData.paymentMethodsResponse, 
```

However, in theory we can change this but not sure, if we then have all the necessary data for the web component call ?

Reply:
@AndreasA 
1.) & 2.) The payment method `type` in the `/paymentMethods` response is unique to each PM, and it does not have to be an array. The response for credit cards would be:
```json
{
  "brands": [
    "visa",
    "mc"
  ],
  "details": [
    {
      "key": "encryptedCardNumber",
      "type": "cardToken"
    },
    {
      "key": "encryptedSecurityCode",
      "type": "cardToken"
    },
    {
      "key": "encryptedExpiryMonth",
      "type": "cardToken"
    },
    {
      "key": "encryptedExpiryYear",
      "type": "cardToken"
    },
    {
      "key": "holderName",
      "optional": true,
      "type": "text"
    }
  ],
  "name": "Credit Card",
  "type": "scheme"
}
```
3.)  However, you made a valid point about the [web component](https://docs.adyen.com/online-payments/components-web#step-2-add-components): according to the docs, the full response from `/paymentMethods` should be used in the configuration. With the current implementation as shown above, `paymentMethod.extensions.adyenData.paymentMethodsResponse.paymentMethods` only returns a single PM, moreover it returns an object instead of an array. 
We can refactor `paymentMethod.extensions.adyenData.paymentMethodsResponse` to only return a single object for each PM type but in the end the client side would still have to make an additional `/paymentMethods` call to get the full response and initiate the component.
The extension `paymentMethod.extensions.adyenData.paymentMethodsResponse` can still be useful when looping though each PM to create and mount the payment method.

Reply:
@peterojo Shouldn't it be enought to provide ALL payment methods for that one payment method in the correct format because anything else is actually really weird. I don't want to be able to switch to Klarna in the web component if I selected credit card and so far it is working fine this way in our implementation.

However, if that is the case, then a separate call the the payment methods response route will be necessary because the payment methods route only returns an array as far as I know.

Reply:
@peterojo Ah. I might have forgotten the following in my code implementaiton:

```
    $paymentMethodsResponse['paymentMethods'] = array_values(array_filter(
            $paymentMethodsResponse['paymentMethods'],
            function ($value) use ($type) {
                return ($value['type'] ?? null) == $type;
            }
        ));
```

it should be like this I think.

Reply:
@peterojo If this doesn't work. I guess we need to remove the extension for payment method again and just always call the full payment methods response controller. 

However, personally I think it really does not make sense to retrieve the payment methods response from adyen to pass it on to adyen again with no modifications whatsoever. That is kind of weird :smile:

Reply:
Otherwise we can also remove the extension and let the client side filter each paymentMethod object from the full response (like we are doing in the plugin)

Reply:
@peterojo Ah. OK. might have misunderstood some parts of the required Adyen web component implementation. However, if I understand your response correctly, the extension should still be used and if it only provides the type but I think it should also provide the payment method config to move as much logic as possible into the Adyen plugin.

- so the full payment methods response as is needed for initialization is provided using the custom route.
- The extension should provide the `details´  for the first payment method and the type which is retrieved from the payment method response be filtering out corresponding data. Before mounting this configuration might be slightly modified to e.g. include the `onChange` handler.

In this case I am all for your change to use `null` or the corresponding object because I would like to remove as much logic as possible from a PWA.

So in your example above:

**Step 1: Initialize adyen checkout**

```
const ADYEN_CHECKOUT_CONFIG = {
    locale,
    clientKey,
    environment,
    showPayButton: false,
    hasHolderName: true,
    paymentMethodsResponse,
    onAdditionalDetails
};
const adyenCheckout = new AdyenCheckout(ADYEN_CHECKOUT_CONFIG);
````

**Step 2: Initialize payment method***
```
adyenCheckout.create(paymentMethod.extensions.adyenData.type, paymentMethod.extensions.adyenData.paymentMethodConfig).mount('#payment-method-container');
```

correct? In wich case you definitely should adjust it to not provide an array but just what is needed.

In this example `paymentMethod.extensions.adyenData.paymentMethodConfig` would be the corresponding payment method config part.


One question which is unrlated to the PR: 
- Can I use the same `AdyenCheckout` object for multiple payment methods or should I create a separate object per payment method. Currently, I create one per payment method but I am not sure this is actually necessary?

Reply:
@AndreasA yes now you understand exactly what I am saying, also I agree with renaming the extension property to `paymentMethodConfig`.
> Can I use the same AdyenCheckout object for multiple payment methods or should I create a separate object per payment method.
Yes you can reuse the same AdyenCheckout object, this is what we do in the plugin as well.

Reply:
@peterojo Yeah.  Then go ahead and adjust the code and rename the property. It makes way more sense that way. We are not using it that way in our project anyway, as we currently have a working solution and did not update yet. Will probably happen in the near futurre though, so it would be nice if the corresponding change is in it then with an official release tag :smile: 

Regarding reusing AdyenCheckout object: Great, so even if different methods are used, the same adyen checkout object is used. Now it also makes more sense to pass on the full payment methods response object because for multiple payment methods the same response object is used. 

Might be nice, if the adyen web component documentation is made a bit clearer on these things as it might help future developers but it is not a must have.